dawn/src/dawn/AST/Offsets.h

```diff
@@ -0,0 +1,164 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+//
+#ifndef DAWN_AST_OFFSET_H
+#define DAWN_AST_OFFSET_H
+
+#include "dawn/Support/Unreachable.h"
+
+#include <memory>
+#include <vector>
+
+namespace dawn::ast {
+
+struct cartesian_ {};
+static constexpr cartesian_ cartesian;
+
+class HorizontalOffsetImpl {
+public:
+  virtual ~HorizontalOffsetImpl() = default;
+  HorizontalOffsetImpl* clone() const { return cloneImpl(); }
+
+  bool operator==(HorizontalOffsetImpl const& other) const { return equalsImpl(other); }
+
+  HorizontalOffsetImpl& operator+=(HorizontalOffsetImpl const& other) { return addImpl(other); }
+
+  bool isZero() const { return isZeroImpl(); }
+
+protected:
+  virtual HorizontalOffsetImpl* cloneImpl() const = 0;
+  virtual bool equalsImpl(HorizontalOffsetImpl const&) const = 0;
+  virtual HorizontalOffsetImpl& addImpl(HorizontalOffsetImpl const&) = 0;
+  virtual bool isZeroImpl() const = 0;
+};
+
+class CartesianOffset : public HorizontalOffsetImpl {
+public:
+  CartesianOffset(int iOffset, int jOffset) : horizontalOffset_({iOffset, jOffset}) {}
+  explicit CartesianOffset(std::array<int, 2> const& offsets) : horizontalOffset_(offsets) {}
+  CartesianOffset() = default;
+
+  int offsetI() const { return horizontalOffset_[0]; }
+  int offsetJ() const { return horizontalOffset_[1]; }
+
+protected:
+  HorizontalOffsetImpl* cloneImpl() const override {
```

Comment:
transfer of ownership, return the unique_ptr already here.

Reply:
I am not 100% sure what's the most idiomatic solution. The reason I did it this way is because ownership is now fully done in `HorizontalOffset` now: All `unique_ptr`s are created and destroyed in there. If clone returns a `unique_ptr`, some allocations happen in `clone` and some others happen in `HorizontalOffset`. I see advantages and disadvantages in both, not sure what is best. What do you think?

Reply:
done